### PR TITLE
ENH: Add ``axis`` kwarg to np.tile

### LINKED ISSUE #8879: ENH: Add axis argument to tile
Currently `tile` doesn't have an `axis` argument. Considering `repeat` and `concatenate`, which share similarities to `tile`, do take `axis` arguments, it doesn't seem unreasonable to have a similar argument for `tile`. Not to mention this would avoid some unnecessary fiddling with arrays to affect the intended dimension. If there were to be an `axis` argument for `tile`, it could be added with a default of `-1`. Thus this would preserve existing behavior while allowing for more flexibility with ease.

### PR DESCRIPTION
**Pull request by Evidlo**
Adds a keyword argument `axis` allowing to specify which axes to tile:

``` python
>>> A = np.arange(3)[None, :]
>>> A
[[0 1 2]]
>>> np.tile(A, (1, 2))
[[0 1 2 0 1 2]]
>>> np.tile(A, 2, axis=-1)
[[0 1 2 0 1 2]]
>>> np.tile(A, (2, 2))
[[0 1 2 0 1 2]
 [0 1 2 0 1 2]]
>>> np.tile(A, (2, 2), axis=(0, 1))
[[0 1 2 0 1 2]
 [0 1 2 0 1 2]]
```

fixes https://github.com/numpy/numpy/issues/8879

This is my first PR to numpy, so let me know if there are any issues.

### CODE REVIEW

File: numpy/lib/_shape_base_impl.py [hunk 1]
```diff
@@ -1260,17 +1274,30 @@ def tile(A, reps):
...
         # have no data there is no risk of an inadvertent overwrite.
-        c = _nx.array(A, copy=False, subok=True, ndmin=d)
-    if (d < c.ndim):
-        tup = (1,)*(c.ndim-d) + tup
+        c = _nx.array(A, copy=False, subok=True, ndmin=len(tup))
+
+    if isinstance(axis, int):
+        if not isinstance(reps, int):
+            raise ValueError("`reps` should be an integer")
+        axis, reps = (axis,), (reps,)
```
Comment (eendebakpt) [inline numpy/lib/_shape_base_impl.py:1289]: This assumes `reps` is an integer, but it could have been passes as a tuple. (The docstring for reps suggests this as it states "array_like")

**Review by eendebakpt — COMMENTED** [review]
The implementation looks good. But I am not sure this functionally should be in numpy. The pr adds some convenience in the input arguments, but this can also be done with a small wrapper around the `np.tile`

File: numpy/lib/_shape_base_impl.py (hunk 1 above)
Comment (Evidlo) [reply numpy/lib/_shape_base_impl.py:1289]: I've updated the docstring for `reps` to include int.

I included this check because `np.tile(A, (2, 2), axis=0)` is self contradictory.

**Comment by Evidlo** [comment]
> The implementation looks good. But I am not sure this functionally should be in numpy. The pr adds some convenience in the input arguments, but this can also be done with a small wrapper around the np.tile

I added this feature to have parity with `np.sum` and `np.roll` (there may be others) which both accept an `axis` kwarg which can be a tuple of ints.

**Comment by Evidlo** [comment]
Any news on this?  Still think this is worth considering due to the principle of Least Astonishment.

File: numpy/lib/_shape_base_impl.py [hunk 2]
```diff
@@ -1185,12 +1185,12 @@ def kron(a, b):
     return result if not is_any_mat else matrix(result, copy=False)
 
 
-def _tile_dispatcher(A, reps):
-    return (A, reps)
+def _tile_dispatcher(A, reps, axis=None):
+    return (A, reps, axis)
```
Comment (mhvk) [inline numpy/lib/_shape_base_impl.py:1189]: `axis` cannot be an array so should not be passed on (since one cannot override the function based on `axis`).

File: numpy/lib/_shape_base_impl.py [hunk 3]
```diff
@@ -1241,16 +1244,29 @@ def tile(A, reps):
 
     >>> b = np.array([[1, 2], [3, 4]])
     >>> np.tile(b, 2)
+    array([[1, 2, 1, 2],
+           [3, 4, 3, 4]])
+    >>> np.tile(b, 2, axis=-1)
```
Comment (mhvk) [inline numpy/lib/_shape_base_impl.py:1249]: I would give your example with `axis=0` here instead, as that immediately shows the benefit.

And then perhaps show a new example with `np.tile(b, (2, 2))`

File: numpy/lib/_shape_base_impl.py [hunk 4]
```diff
@@ -1241,16 +1244,29 @@ def tile(A, reps):
...
+           [3, 4]])
+    >>> np.tile(b, 2, axis=0))
     array([[1, 2],
            [3, 4],
            [1, 2],
            [3, 4]])
 
     >>> c = np.array([1,2,3,4])
-    >>> np.tile(c,(4,1))
+    >>> np.tile(c, (4, 1))
```
Comment (mhvk) [inline numpy/lib/_shape_base_impl.py:1264]: Maybe just replace this example with `np.tile(c, 4, axis=0)` - which really is clearer!

File: numpy/lib/_shape_base_impl.py [hunk 5]
```diff
@@ -1241,16 +1244,29 @@ def tile(A, reps):
...
+    array([[1, 2, 3, 4],
+           [1, 2, 3, 4],
+           [1, 2, 3, 4],
+           [1, 2, 3, 4]])
+    >>> np.tile(c, (4, 1), axis=(0, 1))
```
Comment (mhvk) [inline numpy/lib/_shape_base_impl.py:1269]: And then remove this one.

File: numpy/lib/_shape_base_impl.py [hunk 6]
```diff
@@ -1260,17 +1276,32 @@ def tile(A, reps):
...
+        return _nx.array(A, copy=True, subok=True, ndmin=len(tup))
     else:
         # Note that no copy of zero-sized arrays is made. However since they
         # have no data there is no risk of an inadvertent overwrite.
-        c = _nx.array(A, copy=False, subok=True, ndmin=d)
-    if (d < c.ndim):
-        tup = (1,)*(c.ndim-d) + tup
+        c = _nx.array(A, copy=False, subok=True, ndmin=len(tup))
+
+    if isinstance(axis, int):
```
Comment (mhvk) [inline numpy/lib/_shape_base_impl.py:1288]: Why not just make `axis` a tuple and compare lengths after? I.e., something like (keeping the `d` from above - do feel free to rename it...):
```
if axis is None:
    axis = tuple(range(-d, 0))
else:
    if not isinstance(axis, tuple):
        axis = (axis,)
    elif len(set(axis)) != len(axis):
        raise ValueError("duplicates in `axis`.")
    if len(axis) != d:
        raise ValueError("`reps` and `axis` should have the same length")

tup = tuple((tup[axis.index(i)] if i in axis else 1) for i in range(c.ndim))
shape_out = ...
```
(note: I didn't test this but hopefully it is correct...)

File: numpy/lib/_shape_base_impl.py [hunk 7]
```diff
@@ -1260,17 +1276,32 @@ def tile(A, reps):
         tup = tuple(reps)
     except TypeError:
         tup = (reps,)
-    d = len(tup)
```
Comment (mhvk) [inline numpy/lib/_shape_base_impl.py:1263]: Please do keep `d` here - it is useful below too.

File: numpy/lib/_shape_base_impl.py [hunk 8]
```diff
@@ -1260,17 +1276,32 @@ def tile(A, reps):
...
         # numpy array and the repetitions are 1 in all dimensions
-        return _nx.array(A, copy=True, subok=True, ndmin=d)
+        return _nx.array(A, copy=True, subok=True, ndmin=len(tup))
     else:
         # Note that no copy of zero-sized arrays is made. However since they
         # have no data there is no risk of an inadvertent overwrite.
-        c = _nx.array(A, copy=False, subok=True, ndmin=d)
-    if (d < c.ndim):
-        tup = (1,)*(c.ndim-d) + tup
+        c = _nx.array(A, copy=False, subok=True, ndmin=len(tup))
```
Comment (mhvk) [inline numpy/lib/_shape_base_impl.py:1286]: One thing I somewhat dislike about this is that axis is now an index into this new array, whose dimensions may have changed if `len(tup) > A.ndim`. Maybe this should be `ndmin=(len(tup) if axis is None else 0)`.

(If we go for that, that needs a test too!)

File: numpy/lib/_shape_base_impl.py (hunk 3 above)
Comment (asmeurer) [inline numpy/lib/_shape_base_impl.py:1249]: Maybe I'm missing something, but it seems like every one of your examples is exactly the same as when no `axis` parameter is specified. That doesn't make it very clear what this actually does.

**Review by mhvk — COMMENTED** [review]
I'm not completely sold on this, since in principle one can do everything by just passing in the right `rep`. In this respect, the analogy with `repeat` is not quite right, since there by default the action is on the flattened array (which is not a very good default, but that's too late to fix).

That said, I can see the convenience of doing `np.tile(array, 3, axis=0)` instead of having to write something like `np.tile(array, (3, 1, 1)` for a 3-D array (or having to calculate the `rep` using the number of dimensions. So, overall, +0.5 on the idea.

But some comments on the implementation...

**Comment by mhvk** [comment]
Hmm, I now see @ngoldbaum's reply on the mailing list. He makes a good point that now that there is an API that all array libraries have agreed on, it would be better not to deviate from that unnecessarily.

**Comment by mhvk** [comment]
The further discussion does suggest it is worth proposing it. To be honest, if one were to start from scratch, I think one would have defined `np.tile(array, rep: int, axis: int)` and then expand to both `rep` and `axis` being tuples...

File: numpy/lib/_shape_base_impl.py (hunk 3 above)
Comment (Evidlo) [reply numpy/lib/_shape_base_impl.py:1249]: This specific examples show that the default behavior is equivalent to `axis=-1`

Comment (Evidlo) [reply numpy/lib/_shape_base_impl.py:1249]: I wanted to show what the equivalent call looks like without `axis`.  I thought that it might make it easier to understand its behavior.

Comment (mhvk) [reply numpy/lib/_shape_base_impl.py:1249]: I think the examples should just show the "preferred" way to do things, not every possible way. And I think if we go for adding the `axis` argument, the preferred way would be to use an integer for repeat and the axis to indicate along which axis the repeat should be done.

I think it will be important that the examples show the benefits for suggesting this to the array API.

**Comment by jorenham** [comment]
This seems to have stranded, so I'm going to close it. We can re-open if you still want to work on this @Evidlo

**Comment by Evidlo** [comment]
What are the next steps here?

**Comment by jorenham** [comment]
> What are the next steps here?

I take it you haven't given up on this then? Because if so, then rebasing, resolving the merge conflcits, and addressing the review comments would be a good start :)

**Comment by Evidlo** [comment]
@jorenham @mhvk One thing I'm trying to grapple with is what should happen when `axis` kwarg is a dimension that doesn't exist in the original array.

`np.tile([1, 2, 3, 4], (1, 4))` returns a 2D array.  Do you think that `np.tile([1, 2, 3, 4], 4, axis=1)` should also return a 2D array, or throw an error?

**Comment by Evidlo** [comment]
I have done a complete refactor because I was finding the code a little hard to follow.  The refactor makes the flow a little more linear in my opinion and has these steps:

1. Cast input to an array (do not expand dims yet)
2. Normalize inputs `reps`, `axis`
3. Compute needed dimensions of input and expand if needed
4. Build out `new_reps` tuple
5. Tile (and copy if needed)

<details><summary><strong>Refactored `np.tile`</strong></summary>

``` python
#!/usr/bin/env pythono

import numpy as np
import numpy._core.numeric as _nx
from numpy.testing import (
    assert_, assert_equal, assert_array_equal, assert_raises, assert_warns
    )

def tile(A, reps, axis=None):
    """
    Construct an array by repeating A the number of times given by reps.

    If `reps` has length ``d``, the result will have dimension of
    ``max(d, A.ndim)``.

    If ``A.ndim < d``, `A` is promoted to be d-dimensional by prepending new
    axes. So a shape (3,) array is promoted to (1, 3) for 2-D replication,
    or shape (1, 1, 3) for 3-D replication. If this is not the desired
    behavior, promote `A` to d-dimensions manually before calling this
    function.

    If ``A.ndim > d``, `reps` is promoted to `A`.ndim by prepending 1's to it.
    Thus for an `A` of shape (2, 3, 4, 5), a `reps` of (2, 2) is treated as
    (1, 1, 2, 2).

    Note : Although tile may be used for broadcasting, it is strongly
    recommended to use numpy's broadcasting operations and functions.

    Parameters
    ----------
    A : array_like
        The input array.
    reps : int or array_like
        The number of repetitions of `A` along each axis.  If array-like,
        length should match length of `axis` tuple
    axis : None or int or tuple of ints, optional
        Axis or axes to tile.  Defaults to first axis

    Returns
    -------
    c : ndarray
        The tiled output array.

    See Also
    --------
    repeat : Repeat elements of an array.
    broadcast_to : Broadcast an array to a new shape
    """
    c = _nx.array(A, copy=False, subok=True)

    # Normalize axis and reps
    if axis is None:
        pass
    elif isinstance(axis, int):
        if not isinstance(reps, int):
            raise ValueError("`reps` should be an integer when `axis` is an integer")
        axis, reps = (axis,), (reps,)
    else:
        axis = tuple(axis)
        if isinstance(reps, int):
            reps = (reps,) * len(axis)
        else:
            if len(reps) != len(axis):
                raise ValueError("`reps` and `axis` should have the same length")

    # also handle any list-like items for reps
    try:
        reps = tuple(reps)
    except TypeError:
        reps = (reps,)

    if axis is not None and len(set(axis)) != len(axis):
        raise ValueError("duplicate axes in `axis`")

    # Expand dimensions if necessary
    d = max(len(reps), c.ndim, (max(axis) + 1) if axis else 0)
    if c.ndim < d:
        c = c.reshape((1,) * (d - c.ndim) + c.shape)

    # Build full reps tuple for all dimensions
    if axis is not None:
        tmp = [1] * c.ndim
        for ax, r in zip(axis, reps):
            tmp[ax] = r
        reps = tuple(tmp)
    elif len(reps) < c.ndim:
        reps = (1,) * (c.ndim - len(reps)) + reps

    # Tile
    shape_out = tuple(s*t for s, t in zip(c.shape, reps))
    n = c.size
    if n > 0:
        for dim_in, nrep in zip(c.shape, reps):
            if nrep != 1:
                c = c.reshape(-1, n).repeat(nrep, 0)
            n //= dim_in

    # Ensure we always return a copy, not a view of the input
    if all(r == 1 for r in reps):
        return c.copy().reshape(shape_out)
    return c.reshape(shape_out)

if __name__ == '__main__':
    print('---- main tests ----')
    a = np.array([0, 1, 2])
    b = [[1, 2], [3, 4]]
    assert_equal(tile(a, 2), [0, 1, 2, 0, 1, 2])
    assert_equal(tile(a, 2, axis=1), [[0, 1, 2, 0, 1, 2]])
    assert_equal(tile(a, (2, 2)), [[0, 1, 2, 0, 1, 2], [0, 1, 2, 0, 1, 2]])
    assert_equal(tile(a, (1, 2)), [[0, 1, 2, 0, 1, 2]])
    assert_equal(tile(b, 2), [[1, 2, 1, 2], [3, 4, 3, 4]])
    assert_equal(tile(b, 2, axis=-1), [[1, 2, 1, 2], [3, 4, 3, 4]])
    assert_equal(tile(b, (2, 1)), [[1, 2], [3, 4], [1, 2], [3, 4]])
    assert_equal(tile(b, 2, axis=0), [[1, 2], [3, 4], [1, 2], [3, 4]])
    assert_equal(tile(b, (2, 2)), [[1, 2, 1, 2], [3, 4, 3, 4],
                                    [1, 2, 1, 2], [3, 4, 3, 4]])
    assert_equal(tile(b, (2, 2), axis=(0, 1)),
                    [[1, 2, 1, 2], [3, 4, 3, 4],
                    [1, 2, 1, 2], [3, 4, 3, 4]])
    assert_equal(tile(b, 2, axis=(0, 1)),
                    [[1, 2, 1, 2], [3, 4, 3, 4],
                    [1, 2, 1, 2], [3, 4, 3, 4]])

    print('---- one dimension ----')
    a = np.arange(5)
    b = tile(a, 1)
    b += 2
    assert_equal(a, np.arange(5))

    print('---- test empty ----')
    a = np.array([[[]]])
    b = np.array([[], []])
    c = tile(b, 2).shape
    d = tile(a, (3, 2, 5)).shape
    assert_equal(c, (2, 0))
    assert_equal(d, (3, 2, 0))

    print('---- test kronecker compare ----')
    from numpy.random import randint

    reps = [(2,), (1, 2), (2, 1), (2, 2), (2, 3, 2), (3, 2)]
    shape = [(3,), (2, 3), (3, 4, 3), (3, 2, 3), (4, 3, 2, 4), (2, 2)]
    for s in shape:
        b = randint(0, 10, size=s)
        for r in reps:
            a = np.ones(r, b.dtype)
            large = tile(b, r)
            klarge = np.kron(a, b)
            assert_equal(large, klarge)
```

</details>

I can add to the PR (and rebase) if this looks good.
